packages/loader/container-loader/src/loader.ts

```diff
@@ -202,10 +203,60 @@ export interface ILoaderProps {
 	readonly protocolHandlerBuilder?: ProtocolHandlerBuilder;
 }
 
+/**
+ * Other props which are used to load a container on top of the basic loader props.
+ * @alpha
+ * @legacy
+ */
+export interface IResolveContainerProps extends ILoaderProps {
+	/**
+	 * The request to resolve the container.
+	 */
+	request: IRequest;
+
+	/**
+	 * Pending local state to be applied to the container.
+	 */
+	pendingLocalState?: string;
+}
+
+/**
+ * Other props which are used to create a detached container on top of the basic loader props.
+ * @alpha
+ * @legacy
+ */
+export interface ICreateDetachedContainerProps extends ILoaderProps {
```

Comment:
I'm not a fan of putting canReconnect/clientDetailsOverride/etc. properties on the "new" `ILoaderProps`.  These are `IContainerCreateProps` and were explicitly split out from `ILoaderProps` as they were settings for the container under creation (as opposed to settings used in executing the load of said container).

I also think that if we're creating new functions, we shouldn't tie ourselves to `ILoaderProps`.  Customers will already have to make a code change to move to these so let's make it nice.  At minimum I'm looking at the deprecated members and thinking this should exclude `options: ILoaderOptions`'s deprecated members.  Maybe no detachedBlobStorage either?  Even if that deprecated property is still needed for detached containers, a "detachedBlobStorage" doesn't make sense to me on `IResolveContainerProps` since we're not detached there.

We should also clarify the `canReconnect` functionality entrypoint.  It's already been unified on `IContainerCreateProps.canReconnect`, but the loader uniquely differs how it is set depending on whether you are creating a detached/rehydrated container vs. loading an existing one, which makes no sense and only persists for compat reasons (which again, the new functions don't need to be concerned with).

Reply:
i think you are confusing the legacy surface `ILoaderProps`  and the internal surface `IContainerCreateProps`. i agree we should completely reorganize how this works, but that seems out of scope for just removing the class, which should make it easier to make further changes, as its much easier to support free functions, rather than a class.

Reply:
I don't think I'm confusing anything here, there's a reason why these new additions were kept separate on the `Loader` API surface.

If we're going to make an ask of customers to move to new APIs, we should be thoughtful about the surface that we're presenting so that we're not immediately going back to them for a second breaking API change - this is exactly the kind of friction that our partners have historically complained about.  I think we can do a lot better than `ILoaderProps` without much effort.  We don't get opportunities like adding brand new replacement APIs that often, especially at the loader layer, so I don't think it's that radical to put some time on it.

Removing the class is not a particularly large win here.  It's already near-stateless, it's on the edge of the API surface, and as you mention we don't have major aspirations for changing the functionality offered.  The choices and mistakes we make get locked in for a prolonged time due to partners' ship schedules.  I see no benefit in rushing to eliminate the class at the expense of being thoughtful in our new offering.

Reply:
I'm all for improving the surface, but we need to balance that with incremental change, such that it is easy for existing consumer to transition. I also don't think there is any too fundamentally difficult changing the loader, it is slow, as the layer ships slowly.

I view the major change as removing the class itself. It is currently used as a property bag which gets passed through the layer as the work to remove the loader from scope was never fully completed, so removing the class will be a hurdle regardless of the shape of the replacement.

When the think about the future of the layer, I think we should fundamentally change how it functions:
- Remove IUrlResolver and IDocumentServiceFactory, and have consumer instantiate a strongly type IDocumentService that gets passed into the container; IDocumentService  will need to support clone for the summarizer, which is similar to a factory, but simpler, as the scope is a single container, not any container.
- Try to remove the code loader or at least externalize code loading. I would try to improve typing when the runtime factory is a fixed type but retain the ability for anonymous access. Also gets us out of the way for things like caching 
- Segment api based on scenario, so support different containers type for attached (load) vs detached (create) 
  - Potentially segment even further, maybe a builder pattern, to construct or load a container

Reply:
@ChumpChief So, when you say we can do better than the ILoaderProps, then beside some of the props within IloaderProps like canReconnect and some other which are already deprecated, what else you are targeting? Is it allowing taking arguments directly rather than in form of IloaderProps in the functions or is it about clearing some of the props and reason about each one of them and unify how some of them are used?

Reply:
I'm not trying to debate the value of free functions versus classes.  In this case I don't see much benefit for the reasons I stated and so I don't see a need to rush shipment, but it's not a negative.

However more importantly, I'm offering concrete suggestions to improve the shape of the new API.  It's bizarre to me that you want to discuss "should I be allowed to give this feedback" rather than the material of the feedback itself.

Since the new APIs are now just tagged alpha (rather than alpha+legacy) I wouldn't be opposed to checking in this initial refactor as a staging step (pending the un-deprecation feedback so partners retain a non-deprecated option in the short term).  But I'd like to see some further thought in this space before we promote it to legacy and advise partners to adopt.

Reply:
@jatgarg yes, I think all of the above - as a rule of thumb I would expect new APIs to avoid deprecated members.  I also think it is probably wise to go ahead and avoid the `ILoaderProps` interface itself (not extend it), such that `ILoaderProps` can be easily deprecated/removed at the same time as `Loader`.

And yes for the non-deprecated members yes I think that these have accumulated organically and have not had sufficient review/thought.  Maybe a number of them ultimately stick around, but we should be intentional in our decision to keep them around.